src/network/prover.rs

```diff
@@ -138,9 +154,194 @@ impl<N: Network, E: Environment> Prover<N, E> {
             let _ = handler.await;
         }
 
+        // Initialize the miner, if the node type is a miner.
+        if E::NODE_TYPE == NodeType::Miner && prover.pool.is_none() {
+            Self::start_miner(tasks, prover.clone(), local_ip).await;
+        }
+
+        // Initialize the prover, if the node type is a prover.
+        if E::NODE_TYPE == NodeType::Prover && prover.pool.is_some() {
+            let prover = prover.clone();
+            task::spawn(async move {
+                loop {
+                    // Sleep for `60` seconds.
+                    tokio::time::sleep(std::time::Duration::from_secs(60)).await;
+
+                    // TODO (howardwu): Check that the prover is connected to the pool before proceeding.
+                    //  Currently we use a sleep function to probabilistically ensure the peer is connected.
+                    prover.send_pool_register().await;
+                }
+            });
+        }
+
+        Ok(prover)
+    }
+
+    /// Returns an instance of the prover router.
+    pub fn router(&self) -> ProverRouter<N> {
+        self.prover_router.clone()
+    }
+
+    /// Returns an instance of the memory pool.
+    pub(crate) fn memory_pool(&self) -> Arc<RwLock<MemoryPool<N>>> {
+        self.memory_pool.clone()
+    }
+
+    /// Returns all coinbase records in storage.
+    pub fn to_coinbase_records(&self) -> Vec<(u32, Record<N>)> {
+        self.state.to_coinbase_records()
+    }
+
+    ///
+    /// Performs the given `request` to the prover.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: ProverRequest<N>) {
+        match request {
+            ProverRequest::PoolRequest(operator_ip, share_difficulty, block_template) => {
+                // Process the pool request message.
+                self.process_pool_request(operator_ip, share_difficulty, block_template).await;
+                // Proceed to register the prover to receive the next block template.
+                self.send_pool_register().await;
+            }
+            ProverRequest::MemoryPoolClear(block) => match block {
+                Some(block) => self.memory_pool.write().await.remove_transactions(block.transactions()),
+                None => *self.memory_pool.write().await = MemoryPool::new(),
+            },
+            ProverRequest::UnconfirmedTransaction(peer_ip, transaction) => {
+                // Ensure the node is not peering.
+                if !self.status.is_peering() {
+                    // Process the unconfirmed transaction.
+                    self.add_unconfirmed_transaction(peer_ip, transaction).await
+                }
+            }
+        }
+    }
+
+    ///
+    /// Sends a `PoolRegister` message to the pool IP address.
+    ///
+    async fn send_pool_register(&self) {
+        if E::NODE_TYPE == NodeType::Prover {
```

Comment:
since this method does something only if the node is a prover, should we have a `warn` here in an `else` branch to avoid it being called by non-provers? or do we expect this to be called as a noop by non-provers?

Reply:
I don't think so, as with the other modules we instantiate everything but basically do nothing in case the incorrect node type is assumed. So we probably shouldn't log anything cause it may just end up causing confusion.